app/app/dialog/[id]/page.tsx

```diff
@@ -1,16 +1,241 @@
+      const restoreBackup = async () => {
+        try{
+          if (!client.connected) await client.connect()
+
+          const userId = (await client.getMe()).id
+          setUserId(userId.toString())
+
+          if(!cloudStorage.getKeys.isAvailable()){
+            throw new Error('Error trying to access cloud storage')
+          }
+          
+          const encryptionPassword = await cloudStorage.getItem('encryption-password')
+          if (!encryptionPassword) {
+            throw new Error('Encryption password not found in cloud storage')
+          }
+          
+          const response = await getFromStoracha(`${backupCid}?format=car`)
+          const car = new Uint8Array(await response.arrayBuffer());
+          console.log('CAR file fetched')
+
+          // decode
+          const { roots } = CAR.decode(car)
+          if (!roots.length) {
+            throw new Error('Missing root block')
+          }
+          const { code } = roots[0].cid
+          if (code !== dagCBOR.code) {
+          throw new Error(`unexpected root CID codec: 0x${code.toString(16)}`)
+          }
+
+          console.log('decoding...')
+          const root = roots[0]
+          const backupRaw = dagCBOR.decode(root.bytes) as BackupModel
+          console.log('Decoded CAR root')
+          // console.log(JSON.stringify(backupRaw))
+
+          const id = params.id.replace('-', '')
+          const dialogCid = backupRaw['tg-miniapp-backup@0.0.1'].dialogs[id].toString()
+          const dialogResult = await getFromStoracha(dialogCid)
+          const encryptedDialogData = new Uint8Array(await dialogResult.arrayBuffer())
+
+          const decryptedDialogData = dagCBOR.decode(
+            await Crypto.decryptContent(encryptedDialogData, encryptionPassword)
+          ) as DialogData
+          
+          console.log(JSON.stringify(decryptedDialogData))
+       
+          // Restore messages and entities
+          const restoredMessages: MessageData[] = [];
+          for (const messageLink of decryptedDialogData.messages) {
+            const messagesResult = await getFromStoracha(messageLink.toString())
+            const encryptedMessageData = new Uint8Array(await messagesResult.arrayBuffer())
+            const decryptedMessageData = dagCBOR.decode(
+              await Crypto.decryptContent(encryptedMessageData, encryptionPassword)
+            ) as MessageData[]
+            restoredMessages.push(...decryptedMessageData)
+          }
+    
+          const restoredEntities: Record<string, EntityData> = {}
+          const entitiesResult = await getFromStoracha(decryptedDialogData.entities.toString())
+          const encryptedEntitiesData = new Uint8Array(await entitiesResult.arrayBuffer())
+          const decryptedEntitiesData = dagCBOR.decode(
+            await Crypto.decryptContent(encryptedEntitiesData, encryptionPassword)
+          ) as Record<string, EntityData>
+          Object.assign(restoredEntities, decryptedEntitiesData)
```

Comment:
Consider pulling this logic into the `lib` directory and simply calling it from the UI. Better still use the backup provider and have the provider call the library.